src/modules/fw_pos_control_l1/FixedwingPositionControl.cpp

```diff
@@ -833,6 +836,15 @@ FixedwingPositionControl::control_position(const Vector2f &curr_pos, const Vecto
 
 			float alt_sp = pos_sp_curr.alt;
 
+			if (pos_sp_next.type == position_setpoint_s::SETPOINT_TYPE_LAND && pos_sp_next.valid) {
+				// We're in a loiter directly before a landing WP. Enable our landing configuration (flaps,
+				// landing airspeed and potentially tighter throttle control) already such that we don't
+				// have to do this switch (which can cause significant altitude errors) close to the ground.
+				_tecs.set_time_const_throt(_parameters.land_throtTC_scale * _parameters.time_const_throt);
```

Comment:
Maybe we should be doing this within TECS entirely? TECS internally already has modes which could be expanded as needed for LANDING. We shouldn't have to be continuously setting TECS parameters outside of normal parameter changes.

Reply:
> Maybe we should be doing this within TECS entirely?

That was my thought as well.

Reply:
With "modes" you mean these guys https://github.com/PX4/ecl/blob/1a1106878d6ab1cc48565bef9b25b07d626c5d0b/tecs/tecs.h#L94 ? Yes, this does not contain any LANDING state yet, so that would need to be extended, and i did not want to change @priseborough 's TECS too much here. I agree that if one changes multiple parameters for the landing then TECS should handle this, but i think for a single parameter the current solution would work.